File: documentation/user/en/use/api/example/imperative-schema-definition.cs
evita.UpdateCatalog(
	"evita",
	session => {
		session.DefineEntitySchema("Product")
			/* all is strictly verified but associated data
			   and references can be added on the fly */
			.VerifySchemaButAllow(
				EvolutionMode.AddingAssociatedData,
				EvolutionMode.AddingReferences
			)
			/* products are not organized in the tree */
			.WithoutHierarchy()
			/* prices are referencing another entity stored in Evita */
			.WithPrice()
			/* en + cs localized attributes and associated data are allowed only */
			.WithLocale(new CultureInfo("en"), new CultureInfo("cs-CZ"))
			/* here we define list of attributes with indexes for search / sort */
			.WithAttribute<string>(
				"code",
				whichIs => whichIs.Unique()
			)
			.WithAttribute<string>(
				"url",
				whichIs => whichIs.Unique().Localized()
			)
			.WithAttribute<string[]>(
				"oldEntityUrls",
				whichIs => whichIs.Filterable().Localized()
			)
			.WithAttribute<string>(
				"name",
				whichIs => whichIs.Filterable().Sortable()
			)
			.WithAttribute<string>(
				"ean",
				whichIs => whichIs.Filterable()
			)
			.WithAttribute<long>(
				"priority",
				whichIs => whichIs.Sortable()
			)
			.WithAttribute<DateTimeRange>(
				"validity",
				whichIs => whichIs.filterable()
			)
			.WithAttribute<decimal>(
				"quantity",
				whichIs => whichIs.Filterable().IndexDecimalPlaces(2)
			)
			.WithAttribute(
				"alias", Boolean.class,
				whichIs => whichIs.Filterable()
			)
			/* here we define set of associated data,
			   that can be stored along with entity */
			.WithAssociatedData<ReferencedFileSet>("referencedFiles")
			.WithAssociatedData<Labels>(
				"labels",
				whichIs => whichIs.Localized()
			)
			/* here we define references that relate to
			   another entities stored in Evita */
			.WithReferenceToEntity(
				"categories",
				"Category",
				Cardinality.ZeroOrMore,
				whichIs =>
					/* we can specify special attributes on relation */
					whichIs.Indexed().WithAttribute<long>(
						"categoryPriority",
						thatIs => thatIs.Sortable()
					)
			)
			/* for faceted references we can compute "counts" */
			.WithReferenceToEntity(
				"brand",
				"Brand",
				Cardinality.ZeroOrMore,
				whichIs => whichIs.Faceted())
			/* references may be also represented be
			   entities unknown to Evita */
			.WithReferenceTo(
				"stock",
				"stock",
				Cardinality.ZeroOrMore,
				whichIs => whichIs.Faceted()
			)
			/* finally apply schema changes */
			.UpdateVia(session);
	}
);

File: documentation/user/en/use/api/example/imperative-schema-definition.java
evita.updateCatalog(
	"evita",
	session -> {

		/* first create stubs of the entity schemas that the product will reference */
		session.defineEntitySchema("Brand");
		session.defineEntitySchema("Category");

		session.defineEntitySchema("Product")
			
			/* all is strictly verified but associated data
			   and references can be added on the fly */
			.verifySchemaButAllow(
				EvolutionMode.ADDING_ASSOCIATED_DATA,
				EvolutionMode.ADDING_REFERENCES
			)
			/* products are not organized in the tree */
			.withoutHierarchy()
			/* prices are referencing another entity stored in Evita */
			.withPrice()
			/* en + cs localized attributes and associated data are allowed only */
			.withLocale(Locale.ENGLISH, new Locale("cs", "CZ"))
			/* here we define list of attributes with indexes for search / sort */
			.withAttribute(
				"code", String.class,
				whichIs -> whichIs.unique()
			)
			.withAttribute(
				"url", String.class,
				whichIs -> whichIs.unique().localized()
			)
			.withAttribute(
				"oldEntityUrls", String[].class,
				whichIs -> whichIs.filterable().localized()
			)
			.withAttribute(
				"name", String.class,
				whichIs -> whichIs.filterable().sortable()
			)
			.withAttribute(
				"ean", String.class,
				whichIs -> whichIs.filterable()
			)
			.withAttribute(
				"priority", Long.class,
				whichIs -> whichIs.sortable()
			)
			.withAttribute(
				"validity", DateTimeRange.class,
				whichIs -> whichIs.filterable()
			)
			.withAttribute(
				"quantity", BigDecimal.class,
				whichIs -> whichIs.filterable().indexDecimalPlaces(2)
			)
			.withAttribute(
				"alias", Boolean.class,
				whichIs -> whichIs.filterable()
			)
			/* here we define set of associated data,
			   that can be stored along with entity */
			.withAssociatedData("referencedFiles", ReferencedFileSet.class)
			.withAssociatedData(
				"labels", Labels.class,
				whichIs -> whichIs.localized()
			)
			/* here we define references that relate to
			   another entities stored in Evita */
			.withReferenceToEntity(
				"categories",
				"Category",
				Cardinality.ZERO_OR_MORE,
				whichIs ->
					/* we can specify special attributes on relation */
					whichIs.indexed().withAttribute(
						"categoryPriority", Long.class,
						thatIs -> thatIs.sortable()
					)
			)
			/* for faceted references we can compute "counts" */
			.withReferenceToEntity(
				"brand",
				"Brand",
				Cardinality.ZERO_OR_ONE,
				whichIs -> whichIs.faceted())
			/* references may be also represented be
			   entities unknown to Evita */
			.withReferenceTo(
				"stock",
				"stock",
				Cardinality.ZERO_OR_MORE,
				whichIs -> whichIs.faceted()
			)
			/* finally apply schema changes */
			.updateVia(session);
	}
);
